### PR TITLE
bsunanda:Run2-hcx58 Expand Hcal DetId replacing Run2-hcx29 (11427) and Run2-hcx48 (12406)

### PR DESCRIPTION
**Pull request by bsunanda**
Remove the call with default parameters and replace comparisons in PFlow, HI and Global RecHit comparison codes

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@davidlange6 Please try to see if this PR can be integrated fast. As you see I tried t modify all instantiations of HcalDetId which affected some 80-90 classes spread over all L2's. It is painful to do this many times. This is done > 3 times.

**Comment by deguio** [comment]
+1

File: DataFormats/HcalDetId/src/HcalDetId.cc [hunk 1]
```diff
@@ -26,9 +33,23 @@ HcalDetId::HcalDetId(const DetId& gen) {
 	 subdet!=HcalTriggerTower && subdet!=HcalOther)) {
       throw cms::Exception("Invalid DetId") << "Cannot initialize HcalDetId from " << std::hex << gen.rawId() << std::dec; 
     }  
+    uint32_t rawid = gen.rawId();
+    if ((rawid&kHcalIdFormat2)==0) {
+      int zsid = (rawid&kHcalZsideMask1)?(1):(-1);
```
Comment (slava77) [inline DataFormats/HcalDetId/src/HcalDetId.cc]: about 5 copy-pastes of essentially the same block of code in the same file is suboptimal.

**Comment by slava77** [comment]
It looks like we can not read old SIM anymore:
here is a plot from jenkins wf 50202
all_oldvsnew_ttbarpuwf50202p0c_calotowerssorted_towermaker__reco_obj_obj_eta
both workflows with pileup are affected (50202 and 25202).

**Comment by slava77** [comment]
... and by now there are merge conflicts here as well.

**Comment by kpedro88** [comment]
@slava77 - @bsunanda and I are looking into a better way of implementing this change using Reflex I/O rules as recommended by @lgray. Let's leave this PR open for now, but we may be able to simplify things quite a bit and avoid the comparison problems.

**Comment by slava77** [comment]
@bsunanda @kpedro88 
Is it possible just to add new bits in the available space?
This means ieta and iphi will be split into two bit fields (new "coarse grain" bits and the old bits). They will need to be unpacked and packed with an extra operation.
If it's possible, the old and new formats will be (much more if not fully) compatible.

**Comment by slava77** [comment]
-1
needs updates to resolve old SIM reading in samples with pileup and to get merge conflict resolution

**Comment by davidlange6** [comment]
closing for now - seems lots of work to do here.

**Pull request closed without merging**